main.go

```diff
@@ -3,117 +3,185 @@ package main
 	"github.com/hashicorp/vault-lambda-extension/vault"
+	"github.com/hashicorp/vault/api"
 )
 
 const (
 	extensionName = "vault-lambda-extension"
 )
 
-var (
-	extensionClient = extension.NewClient(os.Getenv("AWS_LAMBDA_RUNTIME_API"))
-)
-
 func main() {
 	logger := log.New(os.Stderr, fmt.Sprintf("[%s] ", extensionName), log.Ldate|log.Ltime|log.LUTC)
 	ctx, cancel := context.WithCancel(context.Background())
+	extensionClient := extension.NewClient(os.Getenv("AWS_LAMBDA_RUNTIME_API"))
+	_, err := extensionClient.Register(ctx, extensionName)
+	if err != nil {
+		logger.Fatal(err)
+	}
 
-	sigs := make(chan os.Signal, 1)
-	signal.Notify(sigs, syscall.SIGTERM, syscall.SIGINT)
+	threadFinishedCh := make(chan struct{}, 2)
+	srv, err := initialiseExtension(logger, threadFinishedCh)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	shutdownChannel := make(chan struct{})
 	go func() {
-		s := <-sigs
+		interruptChannel := make(chan os.Signal, 1)
+		signal.Notify(interruptChannel, syscall.SIGTERM, syscall.SIGINT)
+		select {
+		case s := <-interruptChannel:
+			logger.Printf("Received %s, exiting\n", s)
+		case <-shutdownChannel:
+			logger.Println("Received shutdown event, exiting")
+		}
+
 		cancel()
-		logger.Println("Received", s)
-		logger.Println("Exiting")
+		if err := srv.Shutdown(context.Background()); err != nil {
+			// Error from closing listeners, or context timeout:
+			logger.Printf("HTTP server shutdown error: %s\n", err)
+		}
+		threadFinishedCh <- struct{}{}
 	}()
 
-	_, err := extensionClient.Register(ctx, extensionName)
-	if err != nil {
-		logger.Fatal(err)
-	}
+	processEvents(ctx, logger, extensionClient)
 
-	initialiseExtension(logger)
+	// Once processEvents returns, signal that it's time to shutdown.
+	shutdownChannel <- struct{}{}
 
-	processEvents(ctx, logger)
+	// Ensure we wait for the HTTP server to gracefully shut down.
+	// One signal from the HTTP server thread and one from the shutdown thread.
+	<-threadFinishedCh
+	<-threadFinishedCh
+	logger.Println("Graceful shutdown complete")
 }
 
-func initialiseExtension(logger *log.Logger) {
+func initialiseExtension(logger *log.Logger, threadFinishedCh chan struct{}) (*http.Server, error) {
 	logger.Println("Initialising")
 
 	vaultAddr := os.Getenv("VAULT_ADDR")
 	vaultAuthRole := os.Getenv("VAULT_AUTH_ROLE")
 	vaultAuthProvider := os.Getenv("VAULT_AUTH_PROVIDER")
-	vaultIamServerId := os.Getenv("VAULT_IAM_SERVER_ID")	// Optional
+	vaultIAMServerID := os.Getenv("VAULT_IAM_SERVER_ID") // Optional
 
-	configuredSecrets, err := config.ParseConfiguredSecrets()
+	if vaultAddr == "" || vaultAuthProvider == "" || vaultAuthRole == "" {
+		return nil, errors.New("missing VAULT_ADDR, VAULT_AUTH_PROVIDER or VAULT_AUTH_ROLE environment variables")
+	}
+
+	client, err := vault.NewClient(logger, vaultAuthRole, vaultAuthProvider, vaultIAMServerID)
 	if err != nil {
-		logger.Fatalf("Failed to parse configured secrets to read: %s", err)
+		return nil, fmt.Errorf("error getting client: %s", err)
```

Comment:
We should use error wrapping here and other `fmt.Errorf` calls where we give an error
```suggestion
		return nil, fmt.Errorf("error getting client: %w", err)
```